atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/CoordinationServiceRecorder.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.backup;
+
+import com.google.common.collect.Iterables;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.backup.api.CompletedBackup;
+import com.palantir.atlasdb.coordination.CoordinationService;
+import com.palantir.atlasdb.coordination.ValueAndBound;
+import com.palantir.atlasdb.internalschema.InternalSchemaMetadata;
+import com.palantir.atlasdb.internalschema.InternalSchemaMetadataState;
+import com.palantir.atlasdb.internalschema.persistence.CoordinationServices;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.impl.CheckAndSetResult;
+import com.palantir.atlasdb.timelock.api.Namespace;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.Optional;
+import java.util.function.Function;
+
+final class CoordinationServiceRecorder {
+    private static final SafeLogger log = SafeLoggerFactory.get(CoordinationServiceRecorder.class);
+
+    private final Function<Namespace, KeyValueService> keyValueServiceFactory;
+    private final BackupPersister backupPersister;
+
+    CoordinationServiceRecorder(
+            Function<Namespace, KeyValueService> keyValueServiceFactory, BackupPersister backupPersister) {
+        this.keyValueServiceFactory = keyValueServiceFactory;
+        this.backupPersister = backupPersister;
+    }
+
+    public void storeFastForwardState(CompletedBackup completedBackup) {
+        Namespace namespace = completedBackup.getNamespace();
+        Optional<InternalSchemaMetadataState> maybeMetadata =
+                fetchSchemaMetadata(namespace, completedBackup.getBackupEndTimestamp());
+
+        maybeMetadata.ifPresentOrElse(
+                metadata -> backupPersister.storeSchemaMetadata(completedBackup.getBackupId(), namespace, metadata),
```

Comment:
we might want to return `Optional<InternalSchemaMetadataState>` here and do the actual storage in `AtlasBackupService`?

Reply:
Might be cleaner, but this is also ok -- hard to tell without seeing it